### PR TITLE
AWS SSM secrets manager implementation on ingestion

### PR DESCRIPTION
**Pull request by nahuelverdugo**
Fixes: #6717

### Describe your changes :
Implemented the AWS SSM based secrets manager, refactored the code, tested the whole flow with backend and ingestion, fixed minor errors, and removed some dead code.

### Type of change :
- [x] Bug fix
- [x] Improvement
- [x] New feature

### Checklist:
- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

### Reviewers
Ingestion: @open-metadata/ingestion

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/secrets/SecretsManager.java [hunk 1]
```diff
@@ -48,24 +49,23 @@ public boolean isLocal() {
...
-    encryptOrDecryptDbtConfigSource(ingestionPipeline, ingestionPipeline.getService().getType(), encrypt);
+    encryptOrDecryptDbtConfigSource(ingestionPipeline, ingestionPipeline.getService(), encrypt);
   }
 
   public void encryptOrDecryptDbtConfigSource(
-      IngestionPipeline ingestionPipeline, String serviceType, boolean encrypt) {
+      IngestionPipeline ingestionPipeline, EntityReference service, boolean encrypt) {
     // DatabaseServiceMetadataPipeline contains dbtConfigSource and must be encrypted
-    if (serviceType.equals(Entity.DATABASE_SERVICE)
+    if (service.getType().equals(Entity.DATABASE_SERVICE)
```
Comment (pmbrull) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/secrets/SecretsManager.java:61]: 👍

File: ingestion/src/metadata/utils/secrets/aws_based_secrets_manager.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                    f"No client implemented for auth provider: [{config.authProvider}]"
+                )
+
+    def retrieve_dbt_source_config(
+        self, source_config: SourceConfig, pipeline_name: str
+    ) -> object:
+        logger.debug(
+            f"Retrieving source_config from {self.provider} secrets' manager for {pipeline_name}"
+        )
+        secret_id = self.build_secret_id("database-metadata-pipeline", pipeline_name)
```
Comment (pmbrull) [inline ingestion/src/metadata/utils/secrets/aws_based_secrets_manager.py:100]: should `"database-metadata-pipeline"` become a constant for all the providers?

Comment (pmbrull) [reply ingestion/src/metadata/utils/secrets/aws_based_secrets_manager.py:100]: I see this as the base for AWS, but then we'll add k8s etc

File: ingestion/src/metadata/utils/secrets/local_secrets_manager.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        service: ServiceWithConnectionType,
+        service_type: str,
+    ) -> ServiceConnection:
+        """
+        The LocalSecretsManager does not modify the ServiceConnection object
+        """
+        logger.debug(
+            f"Retrieving service connection from {self.provider} secrets' manager for {service_type} - {service.name}"
+        )
+        return ServiceConnection(__root__=service.connection)
```
Comment (pmbrull) [inline ingestion/src/metadata/utils/secrets/local_secrets_manager.py:59]: Is it needed to pass the `__root__`? Here you might be good to go returning just `service.connection`

**Review by pmbrull — APPROVED** [review]
Just a few nit comments, LGTM.

As a general comment, we are missing a lot of docstrings in both classes and functions. Whenever we add pylint it can be a bit hard to rework it all together

File: ingestion/src/metadata/utils/secrets/local_secrets_manager.py (hunk 3 above)
Comment (nahuelverdugo) [reply ingestion/src/metadata/utils/secrets/local_secrets_manager.py:59]: `service.connection` is an object that could be `DashboardConnection`, `DatabaseConnection`, etc. So in this way, what I am returning is the generic service connection. `Source.serviceConnection` is expecting a `ServiceConnection`.

Comment (pmbrull) [reply ingestion/src/metadata/utils/secrets/local_secrets_manager.py:59]: cool thanks for explaining, LGTM

**Pull request merged**